### PR TITLE
Object Rest/Spread Properties

### PR DESCRIPTION
**Pull request by sebmarkbage**
This proposal was well received when presented to TC-39 and will continue to get work for ES7. Let's start experimenting with it.

https://gist.github.com/sebmarkbage/aa849c7973cb4452c547

@jeffmo @benjamn

### CODE REVIEW

File: component.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "esprima",
-    "version": "3001.0001.0000-dev-harmony-fb",
+    "version": "4001.0001.0000-dev-harmony-fb",
```
Comment (sebmarkbage) [inline component.json:3]: This was previous out-of-sync with package.json which broke `npm test`. Unrelated change and no version bump.

File: esprima.js [hunk 2]
```diff
@@ -5997,7 +6059,7 @@ parseYieldExpression: true
     }
 
     // Sync with *.json manifests.
-    exports.version = '3001.0001.0000-dev-harmony-fb';
+    exports.version = '4001.0001.0000-dev-harmony-fb';
```
Comment (sebmarkbage) [inline esprima.js:6061]: Same out of sync version. Do I need to bump this?

**Comment by sebmarkbage** [comment]
Recommend ignoring whitespace because of indentation changes: https://github.com/facebook/esprima/pull/22/files?w=0

File: test/harmonytest.js [hunk 3]
```diff
@@ -11722,6 +11722,873 @@ var harmonyTestFixture = {
...
+                    "line": 1,
+                    "column": 17
+                }
+            }
+        }
+    },
+
+    // https://gist.github.com/sebmarkbage/aa849c7973cb4452c547
+    'ES7 Proposal: Spread Properties': {
+        'let z = {...x}': {
```
Comment (jeffmo) [inline test/harmonytest.js:12005]: Can you add a non-declaration variant of this test too?
i.e. `z = {...x}` (no `let`)

File: esprima.js [hunk 4]
```diff
@@ -2580,33 +2603,37 @@ parseYieldExpression: true
         expect('{');
 
         while (!match('}')) {
-            property = parseObjectProperty();
```
Comment (jeffmo) [inline esprima.js:2583]: Is this refactoring necessary? I definitely think the way you've restructured things is cleaner, but I'm slightly concerned about future merge conflicts from upstream. If at all possible, it'd make things easier to merge from upstream if we minimize line moves in favor of line insertions where possible

Comment (sebmarkbage) [inline esprima.js:2583]: If you ignore whitespace in the merge tooling you can see that it's just an indentation. https://github.com/facebook/esprima/pull/22/files?w=0 I don't think I can get away with a less invasive diff without duplicating code. :/

File: test/harmonytest.js (hunk 3 above)
Comment (sebmarkbage) [inline test/harmonytest.js:12005]: Done.

File: test/fbtest.js [hunk 5]
```diff
@@ -2043,6 +2043,277 @@ var fbTestFixture = {
...
+                    "column": 0
+                },
+                "end": {
+                    "line": 1,
+                    "column": 18
+                }
+            }
+        },
+
+        '<div {...props} post="attribute" />': {
```
Comment (jeffmo) [inline test/fbtest.js:2155]: Should `<div attr1="foo" {...otherProps} />` be valid?
(and can you add a test to assert either way?)

File: test/fbtest.js [hunk 6]
```diff
@@ -2043,6 +2043,277 @@ var fbTestFixture = {
...
+                    "line": 1,
+                    "column": 0
+                },
+                "end": {
+                    "line": 1,
+                    "column": 35
+                }
+            }
+        }
     },
```
Comment (jeffmo) [inline test/fbtest.js:2570]: Would also be good to add a test to assert that something like this doesn't work: `<div>stuff</div {...props}>`
And probably also that this doesn't work: `<div {...props}>stuff</div {...props}>`
